File: src/voxam/glulx/glk/wrap.py
"""Word wrapping for text buffer windows.

Buffer windows wrap: the game emits a stream of styled characters
and the display decides where the lines break (Glk: Text Buffer
Windows). Only a display knows its width, so wrapping belongs on
the display side -- but every painted display needs the same
thing, so it lives here rather than being written twice.

Two things make this more than a call to textwrap.wrap. Text
arrives in pieces: a window hands over whatever accumulated since
the last flush, which may stop mid-word, so the wrapper keeps the
unfinished paragraph and folds the next piece into it. And text is
styled: breaking a line has to cut the *segments* that make it up,
not a flat string, or the emphasis moves -- so the breaks are
found in the plain text and the segments sliced to match.

The wrapper also keeps the paragraphs it has been given, which is
what makes a resize exact: display lines are recomputed from the
original text rather than re-broken from lines that already lost
their spaces at the break points.
"""

from collections.abc import Iterable, Sequence
from typing import NamedTuple

Segment = tuple[object, str]
"""A run of text sharing one appearance: (key, text).

The key is whatever the display wants to distinguish runs by, and
is only ever compared for equality here -- a Glk style number for
a terminal, something richer for a display that knows links.
"""


class View(NamedTuple):
    """What a window should be showing at this moment.

    Attributes:
        lines: The display lines to show, oldest first.
        start: The index in the wrapper's lines that these begin
            at, for anything anchored to a line number.
        more: Whether text is waiting that this view could not
            fit -- what a pause prompt announces.
    """

    lines: list[list[Segment]]
    start: int
    more: bool


# How many completed paragraphs to remember. Past this the oldest
# are dropped -- a terminal cannot scroll back to them anyway, and
# a long game would otherwise accumulate its entire transcript.
SCROLLBACK = 2000
_TRIM = 200


def _spans(text: str, width: int) -> list[tuple[int, int]]:
    """Index ranges of text, one per display line.

    Newlines are consumed, as is the space at each break. A word
    wider than the line is cut rather than left to overflow.
    """

    width = max(width, 1)
    spans: list[tuple[int, int]] = []
    position = 0
    length = len(text)

    while True:
        newline = text.find("\n", position)
        end = length if newline < 0 else newline
        start = position

        while end - start > width:
            limit = start + width
            # The break may fall on the character just past the
            # line, since a space there costs nothing to drop.
            point = text.rfind(" ", start, limit + 1)

            if point <= start:
                spans.append((start, limit))
                start = limit
            else:
                spans.append((start, point))
                start = point + 1

        spans.append((start, end))

        if newline < 0:
            return spans

        position = newline + 1


def wrap(text: str, width: int) -> list[str]:
    """Break text into lines of at most width characters."""

    return [text[start:end] for start, end in _spans(text, width)]


def wrap_segments(segments: Sequence[Segment], width: int) -> list[list[Segment]]:
    """Break one styled paragraph into styled display lines."""

    if not segments:
        return [[]]

    text = "".join(chunk for _, chunk in segments)
    starts: list[int] = []
    position = 0

    for _, chunk in segments:
        starts.append(position)
        position += len(chunk)

    lines: list[list[Segment]] = []

    for begin, finish in _spans(text, width):
        line: list[Segment] = []

        for index, (key, chunk) in enumerate(segments):
            at = starts[index]
            to = at + len(chunk)

            if to <= begin or at >= finish:
                continue

            piece = chunk[max(begin, at) - at : min(finish, to) - at]

            if piece:
                line.append((key, piece))

        lines.append(line)

    return lines


def plain(line: Iterable[Segment]) -> str:
    """The text of a display line, without its styling."""

    return "".join(chunk for _, chunk in line)


class Wrapper:
    """Accumulates one window's styled output and wraps it to a width.

    Keeps every paragraph, because a full-screen display repaints
    from scratch and cannot re-ask the window for text it has
    already drained.

    Attributes:
        width: The width lines are currently wrapped to.
        seen: How many display lines the player has been shown.
            Everything before this has had its turn on screen;
            text past it that will not fit in one windowful is
            what a pause prompt is for -- see view.
    """

    def __init__(self, width: int = 80) -> None:
        """Start empty, wrapping to the given width."""

        self.width = max(1, width)
        # Completed paragraphs, and the one still being written.
        self._history: list[list[Segment]] = []
        self._current: list[Segment] = []
        # Wrapped forms of each, recomputed only when they change.
        self._done: list[list[Segment]] | None = []
        self._tail: list[list[Segment]] | None = None
        self.seen = 0

    def add(self, runs: Iterable[Segment]) -> None:
        """Fold more styled output in, continuing the open paragraph."""

        for key, text in runs:
            if not text:
                continue

            pieces = text.split("\n")
            self._extend(key, pieces[0])

            for piece in pieces[1:]:
                self._break_paragraph()
                self._extend(key, piece)

        self._tail = None

    def _extend(self, key: object, text: str) -> None:
        if not text:
            return

        if self._current and self._current[-1][0] == key:
            self._current[-1] = (key, self._current[-1][1] + text)
        else:
            self._current.append((key, text))

    def _break_paragraph(self) -> None:
        self._history.append(self._current)

        if self._done is not None:
            self._done.extend(wrap_segments(self._current, self.width))

        self._current = []

        if len(self._history) > SCROLLBACK + _TRIM:
            # Trimmed in batches, so this is not paid on every line.
            del self._history[:_TRIM]
            self._done = None

    @property
    def lines(self) -> list[list[Segment]]:
        """Every display line, oldest first."""

        if self._done is None:
            self._done = [
                line
                for paragraph in self._history
                for line in wrap_segments(paragraph, self.width)
            ]

        if self._tail is None:
            self._tail = wrap_segments(self._current, self.width)

        return self._done + self._tail

    def preview(self, runs: Sequence[Segment]) -> list[list[Segment]]:
        """The display lines as if runs had been added, without adding.

        A display draws the line the player is typing this way: it
        is part of the layout, but it is not part of the window's
        contents until the game accepts it.
        """

        if not runs:
            return self.lines

        lines = self.lines

        return lines[: len(lines) - len(self._tail or [[]])] + wrap_segments(
            [*self._current, *runs], self.width
        )

    # A window shows a windowful. If the game prints more than that
    # between two chances for the player to read, the excess would
    # scroll past unread, so the display stops and waits. The model
    # is glkterm's lastseenline: seen is the high-water mark of what
    # has been shown, and text beyond it that will not fit is what
    # holds things up.

    def view(self, height: int) -> View:
        """What to show now, where it starts, and whether more waits.

        Calling this *is* the display showing them, so it advances
        seen -- but only when there is nothing left waiting. While
        there is, the view stays put until advance is called, which
        is what makes the pause a pause.
        """

        lines = self.lines

        if height <= 0:
            return View([], 0, False)

        if len(lines) - self.seen <= height:
            # Everything unseen fits: show the newest windowful,
            # and the player has now had the lot. Idempotent, which
            # matters -- a repaint happens on every keystroke.
            self.seen = len(lines)
            start = max(0, len(lines) - height)

            return View(lines[start:], start, False)

        start = self._page_start()

        return View(lines[start : start + self._page(height)], start, True)

    def advance(self, height: int) -> None:
        """The player has read a page; move on to the next.

        Always at least one line further on. In a window one or two
        lines tall the page and the overlap are both a single line,
        and without this the pair cancel out and the prompt never
        clears.
        """

        page = self._page_start() + self._page(height)
        self.seen = min(len(self.lines), max(self.seen + 1, page))

    def catch_up(self) -> None:
        """Treat everything as read, however much of it there is.

        For the moments when pausing would be wrong: a file prompt,
        or a window the game has just cleared.
        """

        self.seen = len(self.lines)

    def _page(self, height: int) -> int:
        """Lines of text per page: the window, less the prompt's line."""

        return max(1, height - 1)

    def _page_start(self) -> int:
        # One line of overlap, so the page break does not read as a
        # gap.
        return max(0, self.seen - 1)

    def resize(self, width: int) -> None:
        """Re-wrap everything for a new width."""

        width = max(1, width)

        if width == self.width:
            return

        self.width = width
        self._done = None
        self._tail = None

    def clear(self) -> None:
        """Forget everything, as a cleared window has."""

        self._history = []
        self._current = []
        self._done = []
        self._tail = None
        self.seen = 0

File: csharp/Voxam.Core/Glulx/Glk/Wrap.cs
using System.Text;

namespace Voxam.Core.Glulx.Glk;

/// <summary>
/// What a run of text is dressed in: the Glk style it was written in,
/// and the hyperlink value it belongs to, zero for none.
///
/// The two travel together so a linked run stays distinct from its
/// plain neighbours all the way to the display (Glk: Hyperlinks).
/// </summary>
/// <param name="Style">The Glk style number.</param>
/// <param name="Link">The link value, or zero.</param>
public readonly record struct Dress(uint Style, uint Link);

/// <summary>A run of text sharing one appearance.</summary>
/// <param name="Key">What the display distinguishes runs by.</param>
/// <param name="Text">The characters themselves.</param>
public readonly record struct Segment(Dress Key, string Text);

/// <summary>What a window should be showing at this moment.</summary>
/// <param name="Lines">The display lines to show, oldest first.</param>
/// <param name="Start">
/// The index in the wrapper's lines that these begin at, for anything
/// anchored to a line number.
/// </param>
/// <param name="More">
/// Whether text is waiting that this view could not fit: what a pause
/// prompt announces.
/// </param>
public readonly record struct View(
    IReadOnlyList<IReadOnlyList<Segment>> Lines, int Start, bool More);

/// <summary>
/// Accumulates one window's styled output and wraps it to a width.
///
/// Buffer windows wrap: the game emits a stream of styled characters
/// and the display decides where the lines break (Glk: Text Buffer
/// Windows). Only a display knows its width, so wrapping belongs on the
/// display side, but every painted display needs the same thing, so it
/// lives here rather than being written twice.
///
/// Two things make this more than a call to a line breaker. Text
/// arrives in pieces: a window hands over whatever accumulated since
/// the last flush, which may stop mid-word, so the wrapper keeps the
/// unfinished paragraph and folds the next piece into it. And text is
/// styled: breaking a line has to cut the segments that make it up, not
/// a flat string, or the emphasis moves, so the breaks are found in the
/// plain text and the segments sliced to match.
///
/// The wrapper also keeps the paragraphs it has been given, which is
/// what makes a resize exact: display lines are recomputed from the
/// original text rather than re-broken from lines that already lost
/// their spaces at the break points.
/// </summary>
public sealed class Wrapper
{
    // How many completed paragraphs to remember. Past this the oldest
    // are dropped: a terminal cannot scroll back to them anyway, and a
    // long game would otherwise accumulate its entire transcript.
    private const int Scrollback = 2000;
    private const int Trim = 200;

    // Completed paragraphs, and the one still being written.
    private readonly List<List<Segment>> _history = [];

    private List<Segment> _current = [];

    // Wrapped forms of each, recomputed only when they change.
    private List<List<Segment>>? _done = [];
    private List<List<Segment>>? _tail;

    /// <summary>Start empty, wrapping to the given width.</summary>
    /// <param name="width">The width lines are wrapped to.</param>
    public Wrapper(int width = 80) => Width = Math.Max(1, width);

    /// <summary>The width lines are currently wrapped to.</summary>
    public int Width { get; private set; }

    /// <summary>
    /// How many display lines the player has been shown. Everything
    /// before this has had its turn on screen; text past it that will
    /// not fit in one windowful is what a pause prompt is for.
    /// </summary>
    public int Seen { get; private set; }

    /// <summary>Every display line, oldest first.</summary>
    public IReadOnlyList<IReadOnlyList<Segment>> Lines => Wrapped();

    /// <summary>The text of a display line, without its styling.</summary>
    /// <param name="line">The line to read.</param>
    public static string Plain(IEnumerable<Segment> line)
    {
        ArgumentNullException.ThrowIfNull(line);

        var text = new StringBuilder();

        foreach (var segment in line)
        {
            text.Append(segment.Text);
        }

        return text.ToString();
    }

    /// <summary>Break one styled paragraph into styled display lines.</summary>
    /// <param name="segments">The paragraph, in order.</param>
    /// <param name="width">The width to break it to.</param>
    public static List<List<Segment>> WrapSegments(IReadOnlyList<Segment> segments, int width)
    {
        ArgumentNullException.ThrowIfNull(segments);

        if (segments.Count == 0)
        {
            return [[]];
        }

        // Every paragraph is measured in code points, so a character
        // above the basic plane counts once toward a line's width
        // however many units hold it.
        var points = new int[segments.Count][];
        var starts = new int[segments.Count];
        var length = 0;

        for (var at = 0; at < segments.Count; at++)
        {
            points[at] = Points(segments[at].Text);
            starts[at] = length;
            length += points[at].Length;
        }

        var whole = new int[length];

        for (var at = 0; at < segments.Count; at++)
        {
            points[at].CopyTo(whole, starts[at]);
        }

        var lines = new List<List<Segment>>();

        foreach (var (begin, finish) in Spans(whole, width))
        {
            var line = new List<Segment>();

            for (var at = 0; at < segments.Count; at++)
            {
                var from = starts[at];
                var to = from + points[at].Length;

                if (to <= begin || from >= finish)
                {
                    continue;
                }

                var piece = Joined(
                    points[at], Math.Max(begin, from) - from, Math.Min(finish, to) - from);

                if (piece.Length > 0)
                {
                    line.Add(new Segment(segments[at].Key, piece));
                }
            }

            lines.Add(line);
        }

        return lines;
    }

    /// <summary>Fold more styled output in, continuing the open paragraph.</summary>
    /// <param name="runs">The runs to add, in order.</param>
    public void Add(IEnumerable<Segment> runs)
    {
        ArgumentNullException.ThrowIfNull(runs);

        foreach (var (key, text) in runs)
        {
            if (text.Length == 0)
            {
                continue;
            }

            var pieces = text.Split('\n');

            Extend(key, pieces[0]);

            for (var at = 1; at < pieces.Length; at++)
            {
                BreakParagraph();
                Extend(key, pieces[at]);
            }
        }

        _tail = null;
    }

    /// <summary>
    /// The display lines as if runs had been added, without adding.
    ///
    /// A display draws the line the player is typing this way: it is
    /// part of the layout, but it is not part of the window's contents
    /// until the game accepts it.
    /// </summary>
    /// <param name="runs">The runs to imagine at the end.</param>
    public IReadOnlyList<IReadOnlyList<Segment>> Preview(IReadOnlyList<Segment> runs)
    {
        ArgumentNullException.ThrowIfNull(runs);

        var lines = Wrapped();

        if (runs.Count == 0)
        {
            return lines;
        }

        // Reading the lines settled the open paragraph, and wrapping one
        // always yields at least the empty line, so there is always a
        // tail here to set aside.
        var kept = lines.Take(lines.Count - _tail!.Count).ToList();

        kept.AddRange(WrapSegments([.. _current, .. runs], Width));

        return kept;
    }

    // A window shows a windowful. If the game prints more than that
    // between two chances for the player to read, the excess would
    // scroll past unread, so the display stops and waits. The model is
    // glkterm's lastseenline: Seen is the high-water mark of what has
    // been shown, and text beyond it that will not fit is what holds
    // things up.

    /// <summary>
    /// What to show now, where it starts, and whether more waits.
    ///
    /// Calling this is the display showing them, so it advances Seen,
    /// but only when there is nothing left waiting. While there is, the
    /// view stays put until Advance is called, which is what makes the
    /// pause a pause.
    /// </summary>
    /// <param name="height">How many lines the window can show.</param>
    public View Show(int height)
    {
        var lines = Wrapped();

        if (height <= 0)
        {
            return new View([], 0, false);
        }

        if (lines.Count - Seen <= height)
        {
            // Everything unseen fits: show the newest windowful, and
            // the player has now had the lot. Idempotent, which matters:
            // a repaint happens on every keystroke.
            Seen = lines.Count;

            var newest = Math.Max(0, lines.Count - height);

            return new View(lines.GetRange(newest, lines.Count - newest), newest, false);
        }

        var start = PageStart();
        var page = Math.Min(Page(height), lines.Count - start);

        return new View(lines.GetRange(start, page), start, true);
    }

    /// <summary>
    /// The player has read a page; move on to the next.
    ///
    /// Always at least one line further on. In a window one or two
    /// lines tall the page and the overlap are both a single line, and
    /// without this the pair cancel out and the prompt never clears.
    /// </summary>
    /// <param name="height">How many lines the window can show.</param>
    public void Advance(int height) =>
        Seen = Math.Min(Wrapped().Count, Math.Max(Seen + 1, PageStart() + Page(height)));

    /// <summary>
    /// Treat everything as read, however much of it there is. For the
    /// moments when pausing would be wrong: a file prompt, or a window
    /// the game has just cleared.
    /// </summary>
    public void CatchUp() => Seen = Wrapped().Count;

    /// <summary>Re-wrap everything for a new width.</summary>
    /// <param name="width">The width to wrap to.</param>
    public void Resize(int width)
    {
        width = Math.Max(1, width);

        if (width == Width)
        {
            return;
        }

        Width = width;
        _done = null;
        _tail = null;
    }

    /// <summary>Forget everything, as a cleared window has.</summary>
    public void Clear()
    {
        _history.Clear();
        _current = [];
        _done = [];
        _tail = null;
        Seen = 0;
    }

    /// <summary>
    /// Index ranges of a paragraph, one per display line. Newlines are
    /// consumed, as is the space at each break. A word wider than the
    /// line is cut rather than left to overflow.
    /// </summary>
    private static List<(int Begin, int End)> Spans(int[] text, int width)
    {
        width = Math.Max(width, 1);

        var spans = new List<(int, int)>();
        var position = 0;

        while (true)
        {
            var newline = IndexOf(text, '\n', position);
            var end = newline < 0 ? text.Length : newline;
            var start = position;

            while (end - start > width)
            {
                var limit = start + width;
                // The break may fall on the character just past the
                // line, since a space there costs nothing to drop.
                var point = LastIndexOf(text, ' ', start, Math.Min(limit + 1, text.Length));

                if (point <= start)
                {
                    spans.Add((start, limit));
                    start = limit;
                }
                else
                {
                    spans.Add((start, point));
                    start = point + 1;
                }
            }

            spans.Add((start, end));

            if (newline < 0)
            {
                return spans;
            }

            position = newline + 1;
        }
    }

    private static int IndexOf(int[] text, int wanted, int from)
    {
        for (var at = from; at < text.Length; at++)
        {
            if (text[at] == wanted)
            {
                return at;
            }
        }

        return -1;
    }

    private static int LastIndexOf(int[] text, int wanted, int from, int before)
    {
        for (var at = before - 1; at >= from; at--)
        {
            if (text[at] == wanted)
            {
                return at;
            }
        }

        return -1;
    }

    private static int[] Points(string text)
    {
        var points = new List<int>(text.Length);

        foreach (var rune in text.EnumerateRunes())
        {
            points.Add(rune.Value);
        }

        return [.. points];
    }

    private static string Joined(int[] points, int from, int to)
    {
        var text = new StringBuilder();

        for (var at = from; at < to; at++)
        {
            text.Append(new Rune(points[at]));
        }

        return text.ToString();
    }

    /// <summary>Lines of text per page: the window, less the prompt's line.</summary>
    private static int Page(int height) => Math.Max(1, height - 1);

    // One line of overlap, so the page break does not read as a gap.
    private int PageStart() => Math.Max(0, Seen - 1);

    private List<List<Segment>> Wrapped()
    {
        if (_done is null)
        {
            _done = [];

            foreach (var paragraph in _history)
            {
                _done.AddRange(WrapSegments(paragraph, Width));
            }
        }

        _tail ??= WrapSegments(_current, Width);

        return [.. _done, .. _tail];
    }

    private void Extend(Dress key, string text)
    {
        if (text.Length == 0)
        {
            return;
        }

        if (_current.Count > 0 && _current[^1].Key == key)
        {
            _current[^1] = new Segment(key, _current[^1].Text + text);
        }
        else
        {
            _current.Add(new Segment(key, text));
        }
    }

    private void BreakParagraph()
    {
        _history.Add(_current);
        _done?.AddRange(WrapSegments(_current, Width));
        _current = [];

        if (_history.Count > Scrollback + Trim)
        {
            // Trimmed in batches, so this is not paid on every line.
            _history.RemoveRange(0, Trim);
            _done = null;
        }
    }
}
